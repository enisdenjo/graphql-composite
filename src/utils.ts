import {
  ASTNode,
  DocumentNode,
  FragmentDefinitionNode,
  Kind,
  OperationDefinitionNode,
  SelectionNode,
} from 'graphql';

export function isRecord(val: unknown): val is Record<string, unknown> {
  return val != null && typeof val === 'object' && !Array.isArray(val);
}

/**
 * Flattens fragment definitions to inline fragments of the query.
 * Making sure the gather planning can just use the `Field` visitor and
 * get proper order for the plan.
 *
 * We intentionally dont flatten fragments into just fields because
 * of interfaces/unions support.
 */
export function flattenFragments(doc: DocumentNode): DocumentNode {
  function flattenSelectionNodeIntoSelections(
    node: SelectionNode,
    dest: SelectionNode[],
  ) {
    switch (node.kind) {
      case Kind.FIELD:
      case Kind.INLINE_FRAGMENT: {
        // we leave inline fragments as is because their types are important to interfaces/unions
        const selections: SelectionNode[] = [];
        dest.push({
          ...node,
          selectionSet: {
            kind: Kind.SELECTION_SET,
            selections,
          },
        });
        for (const sel of node.selectionSet?.selections || []) {
          flattenSelectionNodeIntoSelections(sel, selections);
        }
        break;
      }
      case Kind.FRAGMENT_SPREAD: {
        const frag = doc.definitions.find(
          (d) =>
            d.kind === Kind.FRAGMENT_DEFINITION &&
            d.name.value === node.name.value,
        ) as FragmentDefinitionNode | undefined;
        if (!frag) {
          throw new Error(
            `Fragment definition for "${node.name.value}" not found`,
          );
        }
        // we convert a fragment definition to an inline fragment to retain the type information
        const selections: SelectionNode[] = [];
        dest.push({
          kind: Kind.INLINE_FRAGMENT,
          typeCondition: frag.typeCondition,
          selectionSet: {
            kind: Kind.SELECTION_SET,
            selections,
          },
        });
        for (const sel of frag.selectionSet.selections) {
          flattenSelectionNodeIntoSelections(sel, selections);
        }
        break;
      }
    }
  }

  const definitions: OperationDefinitionNode[] = [];
  for (const def of doc.definitions.filter(isOperationDefinitionNode)) {
    const selections: SelectionNode[] = [];
    definitions.push({
      ...def,
      selectionSet: {
        kind: Kind.SELECTION_SET,
        selections,
      },
    });
    for (const sel of def.selectionSet.selections) {
      flattenSelectionNodeIntoSelections(sel, selections);
    }
  }
  return {
    kind: Kind.DOCUMENT,
    definitions,
  };
}

function isOperationDefinitionNode(
  node: ASTNode,
): node is OperationDefinitionNode {
  return node.kind === Kind.OPERATION_DEFINITION;
}
