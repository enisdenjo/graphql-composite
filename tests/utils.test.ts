import { parse, print } from 'graphql';
import { describe, expect, it } from 'vitest';
import { flattenFragments } from '../src/utils.js';

describe('flattenFragments', () => {
  it.each([
    {
      name: 'BasicInlineFragment',
      document: parse(/* GraphQL */ `
        {
          storefront {
            products {
              ... on Product {
                upc
                name
              }
            }
          }
        }
      `),
    },
    {
      name: 'NestedInlineFragment',
      document: parse(/* GraphQL */ `
        {
          storefront {
            products {
              ... on Product {
                name
                upc
                manufacturer {
                  ... on Manufacturer {
                    name
                  }
                  id
                }
              }
            }
          }
        }
      `),
    },
    {
      name: 'BasicFragmentDefinition',
      document: parse(/* GraphQL */ `
        {
          storefront {
            products {
              ...P
            }
          }
        }
        fragment P on Product {
          name
          upc
        }
      `),
    },
    {
      name: 'NestedFragmentDefinition',
      document: parse(/* GraphQL */ `
        {
          storefront {
            products {
              ...P
            }
          }
        }
        fragment P on Product {
          name
          upc
          manufacturer {
            ...M
            groundedAt
          }
        }
        fragment M on Manufacturer {
          id
          name
        }
      `),
    },
    {
      name: 'AliasesFragmentDefinition',
      document: parse(/* GraphQL */ `
        {
          storefront {
            pro: products {
              ...P
            }
          }
        }
        fragment P on Product {
          name
          id: upc
        }
      `),
    },
    {
      name: 'DirectivesFragmentDefinition',
      document: parse(/* GraphQL */ `
        {
          storefront {
            products {
              ...P
            }
          }
        }
        fragment P on Product {
          name
          upc @skip(if: false)
          manufacturer @include(if: true) {
            groundedAt
          }
        }
      `),
    },
    {
      name: 'DuplicateFieldsFragmentDefinition',
      document: parse(/* GraphQL */ `
        {
          storefront {
            products {
              ...P1
              ...P2
            }
          }
        }
        fragment P1 on Product {
          name
          upc
          manufacturer {
            name
            groundedAt
          }
        }
        fragment P2 on Product {
          id
          upc
          manufacturer {
            name
          }
        }
      `),
    },
  ])('should flatten fragments in $name', ({ document }) => {
    const flat = flattenFragments(document);
    expect(print(flat)).toMatchSnapshot();
  });
});
