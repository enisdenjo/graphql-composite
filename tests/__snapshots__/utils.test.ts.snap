// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`flattenFragments > should flatten fragments in 'AliasesFragmentDefinition' 1`] = `
"{
  storefront {
    pro: products {
      ... on Product {
        name
        id: upc
      }
    }
  }
}"
`;

exports[`flattenFragments > should flatten fragments in 'BasicFragmentDefinition' 1`] = `
"{
  storefront {
    products {
      ... on Product {
        name
        upc
      }
    }
  }
}"
`;

exports[`flattenFragments > should flatten fragments in 'BasicInlineFragment' 1`] = `
"{
  storefront {
    products {
      ... on Product {
        upc
        name
      }
    }
  }
}"
`;

exports[`flattenFragments > should flatten fragments in 'DirectivesFragmentDefinition' 1`] = `
"{
  storefront {
    products {
      ... on Product {
        name
        upc @skip(if: false)
        manufacturer @include(if: true) {
          groundedAt
        }
      }
    }
  }
}"
`;

exports[`flattenFragments > should flatten fragments in 'DuplicateFieldsFragmentDefinition' 1`] = `
"{
  storefront {
    products {
      ... on Product {
        name
        upc
        manufacturer {
          name
          groundedAt
        }
      }
      ... on Product {
        id
        upc
        manufacturer {
          name
        }
      }
    }
  }
}"
`;

exports[`flattenFragments > should flatten fragments in 'NestedFragmentDefinition' 1`] = `
"{
  storefront {
    products {
      ... on Product {
        name
        upc
        manufacturer {
          ... on Manufacturer {
            id
            name
          }
          groundedAt
        }
      }
    }
  }
}"
`;

exports[`flattenFragments > should flatten fragments in 'NestedInlineFragment' 1`] = `
"{
  storefront {
    products {
      ... on Product {
        name
        upc
        manufacturer {
          ... on Manufacturer {
            name
          }
          id
        }
      }
    }
  }
}"
`;
